contracts/EtomicSwapV2.sol

```diff
@@ -0,0 +1,483 @@
+        address tokenAddress
+    ) external {
+        require(
+            makerPayments[id].state == MakerPaymentState.PaymentSent,
+            "Invalid payment state. Must be PaymentSent"
+        );
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                amount,
+                taker,
+                msg.sender,
+                takerSecretHash,
+                makerSecretHash,
+                tokenAddress
+            )
+        );
+
+        require(
+            paymentHash == makerPayments[id].paymentHash,
+            "Invalid paymentHash"
+        );
+
+        require(
+            block.timestamp >= makerPayments[id].paymentLockTime,
+            "Current timestamp didn't exceed payment refund lock time"
+        );
+
+        makerPayments[id].state = MakerPaymentState.MakerRefunded;
+
+        emit MakerPaymentRefundedTimelock(id);
+
+        if (tokenAddress == address(0)) {
+            payable(msg.sender).transfer(amount);
+        } else {
+            IERC20 token = IERC20(tokenAddress);
+            require(token.transfer(msg.sender, amount));
+        }
+    }
+
+    function refundMakerPaymentSecret(
+        bytes32 id,
+        uint256 amount,
+        address taker,
+        bytes32 takerSecret,
+        bytes20 makerSecretHash,
+        address tokenAddress
+    ) external {
+        require(
+            makerPayments[id].state == MakerPaymentState.PaymentSent,
+            "Invalid payment state. Must be PaymentSent"
+        );
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                amount,
+                taker,
+                msg.sender,
+                ripemd160(abi.encodePacked(sha256(abi.encodePacked(takerSecret)))),
+                makerSecretHash,
+                tokenAddress
+            )
+        );
+
+        require(
+            paymentHash == makerPayments[id].paymentHash,
+            "Invalid paymentHash"
+        );
+
+        makerPayments[id].state = MakerPaymentState.MakerRefunded;
+
+        emit MakerPaymentRefundedSecret(id, takerSecret);
+
+        if (tokenAddress == address(0)) {
+            payable(msg.sender).transfer(amount);
+        } else {
+            IERC20 token = IERC20(tokenAddress);
+            require(token.transfer(msg.sender, amount));
+        }
+    }
+
+    function ethTakerPayment(
+        bytes32 id,
+        uint256 dexFee,
+        address receiver,
+        bytes20 takerSecretHash,
+        bytes20 makerSecretHash,
+        uint32 immediateRefundLockTime,
+        uint32 paymentLockTime
+    ) external payable {
+        require(takerPayments[id].state == TakerPaymentState.Uninitialized, "Taker payment is already initialized");
+        require(receiver != address(0), "Receiver must not be zero address");
+        require(msg.value > 0, "ETH value must be greater than zero");
+        require(msg.value > dexFee, "ETH value must be greater than dex fee");
```

Comment:
This require might not be needed since after Solidity 0.8.0 underflow is detected https://docs.soliditylang.org/en/latest/080-breaking-changes.html

Reply:
Yeah, good note, thanks 🙂

Reply:
> the dexFee might be equal to msg.value

But shouldn't this be ensured by the defi framework implementation instead. I don't think it will make a big difference if we removed this require either way so we can just keep it.